### PR TITLE
Create ring buffer for re-used command lists

### PR DESCRIPTION
**Pull request by raoanag**

### CODE REVIEW

File: onnxruntime/core/providers/dml/DmlExecutionProvider/src/FusedGraphKernel.cpp [hunk 1]
```diff
@@ -13,6 +13,26 @@ namespace Dml
...
+
+            // Bindings from previous executions of a re-used command list
+            mutable std::vector<uint64_t> inputBindingAllocIds;
+            mutable std::vector<uint64_t> outputBindingAllocIds;
+            mutable uint64_t tempBindingAllocId = 0;
+
+            // Fence tracking the status of the command list's last execution, and whether its descriptor heap
+            // can safely be updated.
+            mutable ComPtr<ID3D12Fence> fence;
+            mutable uint64_t completionValue = 0;
```
Comment (PatriceVignola) [inline onnxruntime/core/providers/dml/DmlExecutionProvider/src/FusedGraphKernel.cpp:33]: Are those `mutable` keywords still necessary? `mutable std::deque<std::unique_ptr<ReusedCommandListState>> m_reusedCommandLists;` is already marked as mutable below.

Comment (raoanag) [reply onnxruntime/core/providers/dml/DmlExecutionProvider/src/FusedGraphKernel.cpp:33]: I believe it makes the deque mutable.
If I take it off the build errors when modifying `m_reusedCommandLists` Line 172
`m_reusedCommandLists.push_front(BuildReusableCommandList());`

Comment (yuslepukhin) [reply onnxruntime/core/providers/dml/DmlExecutionProvider/src/FusedGraphKernel.cpp:33]: This is usually the case when a const method attempts to change data. Should the method NOT be const?

Comment (jeffbloo) [reply onnxruntime/core/providers/dml/DmlExecutionProvider/src/FusedGraphKernel.cpp:33]: The EP more generally does not support concurrent execution.  The kernel needs to maintain caches, and those caches are opaque to ORT.  I don't believe there are much better alternatives in this case, and it's following a pre-existing pattern.

**Review by jstoecker — APPROVED** [review]
Approving. This is identical to existing logic (also had mutable fields) and we simply have N groups of reusable command list state instead of 1 now. Like Jeff said, this follows a pre-existing pattern.

**Review by sumitsays — APPROVED** [review]

**Pull request merged**